<a href="https://colab.research.google.com/github/AiswaryaGoriparthi/Aiswarya_INFO5731_Fall2024/blob/main/Goriparthi_Aiswarya_Assignment_01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **INFO5731 Assignment One**

This assignment includes range of programming tasks and data analysis exercises, aiming to enhance your Python skills and proficiency in string manipulation, numerical computations, and data science. Follow the instructions carefully and design well-structured Python programs to address each question.

Late Submission will have a penalty of 10% reduction for each day after the deadline.

# Question 1 (10 points).

The formula that describes the volume V of a hemisphere with radius r is the following:

$ V=\frac{2}{3}\ast\ \pi\ast\ r^3 $

Write a program to calculate the value of V when r is in the range of even numbers from 1-20. Output the result in the following format:

	r	V
	2	…
	4	…
	6	…
	.
	.
	.
	20	…


In [1]:
# Your code here
import math

# Function to calculate the volume of a hemisphere
def hemisphere_vol(r):
    return (2/3) * math.pi * (r**3)

# Output table header
print(f"{'r':<5}{'V':<20}")

# Calculate and print volume for even numbers in the range 1 to 20
for r in range(2, 21, 2):
    V = hemisphere_vol(r)
    print(f"{r:<5}{V:<20.5f}")





r    V                   
2    16.75516            
4    134.04129           
6    452.38934           
8    1072.33029          
10   2094.39510          
12   3619.11474          
14   5747.02016          
16   8578.64234          
18   12214.51224         
20   16755.16082         


# Question 2 (20 points)

(a) Write a Python program to generate a sequence of 12 random numbers falling within the range of 50 to 150.

(b) Arrange the generated numbers in ascending order. Display the sorted list.

(c) Extend the program to arrange the numbers in descending order as well. Present the sorted list accordingly.

(d) Compute and output the average of the entire group of generated numbers.

(e) Enhance the program to calculate and display the average of the first 6 numbers from both the ascending and descending sorted lists.

In [2]:
# Write code for each of the sub parts with proper comments
import random

# (a) Generating a sequence of 12 random numbers between 50 and 150
random_numbers = [random.randint(50, 150) for _ in range(12)]
# Print the random numbers
print("Random numbers:", random_numbers)

# (b) Sort the numbers in ascending order
ascending_numbers = sorted(random_numbers)
print("Sorted numbers (Ascending):", ascending_numbers)

# (c) Sort the numbers in descending order
descending_numbers = sorted(random_numbers, reverse=True)
print("Sorted numbers (Descending):", descending_numbers)

# (d) Compute and output the average of the entire group of generated numbers
tot_average = sum(random_numbers) / len(random_numbers)
print("Average of the entire group:", round(tot_average, 2))

# (e) Calculate and display the average of the first 6 numbers from both the ascending and descending sorted lists
average_first_6_ascending_num = sum(ascending_numbers[:6]) / 6
average_first_6_descending_num = sum(descending_numbers[:6]) / 6

print("Average of the first 6 numbers (Ascending):", round(average_first_6_ascending_num, 2))
print("Average of the first 6 numbers (Descending):", round(average_first_6_descending_num, 2))


Random numbers: [50, 106, 70, 116, 106, 92, 72, 56, 94, 86, 127, 55]
Sorted numbers (Ascending): [50, 55, 56, 70, 72, 86, 92, 94, 106, 106, 116, 127]
Sorted numbers (Descending): [127, 116, 106, 106, 94, 92, 86, 72, 70, 56, 55, 50]
Average of the entire group: 85.83
Average of the first 6 numbers (Ascending): 64.83
Average of the first 6 numbers (Descending): 106.83


# explaining in detail for every section from a to e of 2nd question


In [3]:
#  Executing every section individually
# (a) Write a Python program to generate a sequence of 12 random numbers falling within the range of 50 to 150.
import random

# Generate 12 random numbers between 50 and 150
random_numbers = [random.randint(50, 150) for _ in range(12)]

# Print the random numbers
print("Random numbers:", random_numbers)




Random numbers: [116, 110, 121, 138, 59, 67, 123, 72, 119, 100, 132, 87]


In [4]:
#b) Arrange the generated numbers in ascending order. Display the sorted list.
import random

# Generate 12 random numbers between 50 and 150
random_numbers = [random.randint(50, 150) for _ in range(12)]

# Print the random numbers
print(" The Random numbers:", random_numbers)

# Sort the numbers in ascending order
random_numbers.sort()

# Print the sorted list
print("Numbers after Sorting random numbers:", random_numbers)


 The Random numbers: [105, 114, 55, 118, 123, 56, 75, 116, 132, 133, 143, 119]
Numbers after Sorting random numbers: [55, 56, 75, 105, 114, 116, 118, 119, 123, 132, 133, 143]


In [5]:
# (c) Extend the program to arrange the numbers in descending order as well. Present the sorted list accordingly.
import random

# Generate 12 random numbers between 50 and 150
random_numbers = [random.randint(50, 150) for _ in range(12)]

# Sort the numbers in ascending order
ascending_numbers = sorted(random_numbers)

# Sort the numbers in descending order
descending_numbers = sorted(random_numbers, reverse=True)


# Print the random numbers
print("Random numbers:", random_numbers)

# Print the sorted lists
print("Sorted numbers (Ascending):", ascending_numbers)
print("Sorted numbers (Descending):", descending_numbers)


Random numbers: [122, 60, 91, 65, 146, 128, 130, 135, 105, 91, 98, 145]
Sorted numbers (Ascending): [60, 65, 91, 91, 98, 105, 122, 128, 130, 135, 145, 146]
Sorted numbers (Descending): [146, 145, 135, 130, 128, 122, 105, 98, 91, 91, 65, 60]


In [6]:
# (d) Compute and output the average of the entire group of generated numbers.
import random

# Generate 12 random numbers between 50 and 150
random_numbers = [random.randint(50, 150) for _ in range(12)]

# Sort the numbers in ascending order
ascending_numbers = sorted(random_numbers)

# Sort the numbers in descending order
descending_numbers = sorted(random_numbers, reverse=True)

# Calculate the average of the numbers
average = sum(random_numbers) / len(random_numbers)

# Print the random numbers
print(" The Random numbers:", random_numbers)

# Print the sorted lists and the average
print("Sorted numbers (Ascending):", ascending_numbers)
print("Sorted numbers (Descending):", descending_numbers)
print("Average of the numbers:", round(average, 2))


 The Random numbers: [130, 98, 76, 134, 110, 110, 147, 97, 97, 117, 135, 60]
Sorted numbers (Ascending): [60, 76, 97, 97, 98, 110, 110, 117, 130, 134, 135, 147]
Sorted numbers (Descending): [147, 135, 134, 130, 117, 110, 110, 98, 97, 97, 76, 60]
Average of the numbers: 109.25


In [7]:
# (e) Enhance the program to calculate and display the average of the first 6 numbers from both the ascending and descending sorted lists.
import random

# Generate 12 random numbers between 50 and 150
random_numbers = [random.randint(50, 150) for _ in range(12)]

# Sort the numbers in ascending order
ascending_numbers = sorted(random_numbers)

# Sort the numbers in descending order
descending_numbers = sorted(random_numbers, reverse=True)

# Calculate the average of the entire list
average = sum(random_numbers) / len(random_numbers)

# Calculate the average of the first 6 numbers in ascending order
average_first_6_ascending_order = sum(ascending_numbers[:6]) / 6

# Calculate the average of the first 6 numbers in descending order
average_first_6_descending_order = sum(descending_numbers[:6]) / 6

# Print the random numbers
print("Random numbers:", random_numbers)

# Print the sorted lists and the averages
print("Sorted numbers (Ascending):", ascending_numbers)
print("Sorted numbers (Descending):", descending_numbers)
print("Average of the entire group:", round(average, 2))
print("Average of the first 6 (Ascending):", round(average_first_6_ascending_order, 2))
print("Average of the first 6 (Descending):", round(average_first_6_descending_order, 2))


Random numbers: [116, 103, 143, 139, 74, 71, 147, 125, 55, 97, 145, 61]
Sorted numbers (Ascending): [55, 61, 71, 74, 97, 103, 116, 125, 139, 143, 145, 147]
Sorted numbers (Descending): [147, 145, 143, 139, 125, 116, 103, 97, 74, 71, 61, 55]
Average of the entire group: 106.33
Average of the first 6 (Ascending): 76.83
Average of the first 6 (Descending): 135.83


# Question 3 (20 points).

Create a Python program that encompasses various string manipulation task:

(a) Prompt the user to input two sentences.

(b) Transform each sentence into a list, where each element corresponds to a word in the sentence. For example, if the sentences are "Hello, how are you doing today?" and "This is an example sentence for the program!"

The program should output:
['Hello', ',', 'how', 'are', 'you', 'doing', 'today', '?', 'This', 'is', 'an', 'example', 'sentence', 'for', 'the', 'program', '!']

(c) Compare the two sentences. Print out a list of words that occur in both sentences.

(d) Remove any punctuation from the sentence and print the sanitized sentences. It should remove at least the following characters, but it can remove more: period(“.”), comma (“,”), semicolon (“;”), and colon (“:”).

(e) Count the number of vowels in each of the sentences. Print out each vowel and the occurrences  in the each of the two sentences separately, such as:

Sentence1: a:2, e:2, i:0, etc.
Sentence2: a:1, e:2, i:2, etc.

Design a well-structured Python program to execute these tasks efficiently.

In [8]:
# Your code here

import string

# Function to remove punctuation from a sentence
def remove_punctuations(sentence):
    return sentence.translate(str.maketrans('', '', string.punctuation))

# Function to count vowels in a sentence
def count_vowels(sentence):
    vowels = "aeiouAEIOU"
    count = {v: sentence.count(v) for v in vowels}
    return count

# Function to compare and find common words in two lists
def common_words(list1, list2):
    return list(set(list1) & set(list2))

# (a) Prompt the user to input two sentences
sentence1 = input("Enter the first sentence: ")
sentence2 = input("Enter the second sentence: ")

# (b) Transform each sentence into a list of words
sentence1_list = sentence1.split()
sentence2_list = sentence2.split()

# Output the combined list of words
combined_list = sentence1_list + sentence2_list
print("\nCombined list of words:", combined_list)

# (c) Compare the two sentences and print common words
common = common_words(sentence1_list, sentence2_list)
print("\nCommon words in both sentences:", common)

# (d) Remove punctuation from the sentences

print("\nSentence 1 without punctuation:", remove_punctuations(sentence1))
print("Sentence 2 without punctuation:", remove_punctuations(sentence2))

# (e) Count vowels in each sentence
vowel_count_1 = count_vowels(remove_punctuations(sentence1))
vowel_count_2 = count_vowels(remove_punctuations(sentence2))

# Print vowel counts for each sentence
print("\nVowel count in Sentence 1:")
for vowel, count in vowel_count_1.items():
    print(f"{vowel}: {count}")

print("\nVowel count in Sentence 2:")
for vowel, count in vowel_count_2.items():
    print(f"{vowel}: {count}")




Enter the first sentence: where is veer and Rhea?
Enter the second sentence: veer and Rhea are playing in the garden!

Combined list of words: ['where', 'is', 'veer', 'and', 'Rhea?', 'veer', 'and', 'Rhea', 'are', 'playing', 'in', 'the', 'garden!']

Common words in both sentences: ['veer', 'and']

Sentence 1 without punctuation: where is veer and Rhea
Sentence 2 without punctuation: veer and Rhea are playing in the garden

Vowel count in Sentence 1:
a: 2
e: 5
i: 1
o: 0
u: 0
A: 0
E: 0
I: 0
O: 0
U: 0

Vowel count in Sentence 2:
a: 5
e: 6
i: 2
o: 0
u: 0
A: 0
E: 0
I: 0
O: 0
U: 0


# explaining in detail for every section from a to e of 3rd question

In [9]:
# explaining in detail for every section from a to e
# (a) Prompt the user to input two sentences.
sentence_1 = input("Enter the first sentence: ")
sentence_2 = input("Enter the second sentence: ")

# Display the entered sentences
print("\nFirst Sentence:", sentence_1)
print("Second Sentence:", sentence_2)


Enter the first sentence: where is veer and Rhea?
Enter the second sentence: veer and Rhea are playing in the garden!

First Sentence: where is veer and Rhea?
Second Sentence: veer and Rhea are playing in the garden!


In [10]:
# (b) Transform each sentence into a list of words
sentence1 = input("Enter the first sentence: ")
sentence2 = input("Enter the second sentence: ")

# Split the sentences into lists of words
sentence1_list = sentence1.split()
sentence2_list = sentence2.split()

# Display the lists
print("\nList of words in the first sentence:", sentence1_list)
print("List of words in the second sentence:", sentence2_list)



Enter the first sentence: where is veer and Rhea?
Enter the second sentence: veer and Rhea are playing in the garden!

List of words in the first sentence: ['where', 'is', 'veer', 'and', 'Rhea?']
List of words in the second sentence: ['veer', 'and', 'Rhea', 'are', 'playing', 'in', 'the', 'garden!']


In [11]:
# (c) Compare the two sentences and print common words
def find_common_words(sentence1, sentence2):
    # Split the sentences into words
    sentence1_list = sentence1.split()
    sentence2_list = sentence2.split()

    # Find common words using set intersection
    common_words = list(set(sentence1_list) & set(sentence2_list))

    return common_words

# Prompt the user to input two sentences
sentence1 = input("Enter the first sentence: ")
sentence2 = input("Enter the second sentence: ")

# Find and display common words
common_words = find_common_words(sentence1, sentence2)
print("\nCommon words in both sentences:", common_words)



Enter the first sentence: where is veer and Rhea?
Enter the second sentence: veer and Rhea are playing in the garden!

Common words in both sentences: ['veer', 'and']


In [12]:
import string

# (d) Function to remove punctuation from a sentence
def remove_punctuation(sentence):
    return sentence.translate(str.maketrans('', '', string.punctuation))

# Prompt the user to input two sentences
sentence1 = input("Enter the first sentence: ")
sentence2 = input("Enter the second sentence: ")

# Remove punctuation from both sentences
sentence1_clean = remove_punctuation(sentence1)
sentence2_clean = remove_punctuation(sentence2)

# Display the sanitized sentences
print("\nFirst sentence without punctuation:", sentence1_clean)
print("Second sentence without punctuation:", sentence2_clean)


Enter the first sentence: where is veer and Rhea?
Enter the second sentence: veer and Rhea are playing in the garden!

First sentence without punctuation: where is veer and Rhea
Second sentence without punctuation: veer and Rhea are playing in the garden


In [13]:
#e) Function to count vowels in a sentence
def count_vowels(sentence):
    vowels = "aeiouAEIOU"
    # Create a dictionary to hold the count for each vowel
    vowel_count = {v: sentence.count(v) for v in vowels}
    return vowel_count

# Prompt the user to input two sentences
sentence1 = input("Enter the first sentence: ")
sentence2 = input("Enter the second sentence: ")

# Count the vowels in each sentence
vowel_count_1 = count_vowels(sentence1)
vowel_count_2 = count_vowels(sentence2)

# Display the vowel counts for the first sentence
print("\nVowel count in Sentence 1:")
for vowel in "aeiou":
    print(f"{vowel}: {vowel_count_1.get(vowel, 0)}")

# Display the vowel counts for the second sentence
print("\nVowel count in Sentence 2:")
for vowel in "aeiou":
    print(f"{vowel}: {vowel_count_2.get(vowel, 0)}")


Enter the first sentence: where is veer and Rhea?
Enter the second sentence: veer and Rhea are playing in the garden!

Vowel count in Sentence 1:
a: 2
e: 5
i: 1
o: 0
u: 0

Vowel count in Sentence 2:
a: 5
e: 6
i: 2
o: 0
u: 0


# Question 4 (20 Points)

Design and implement a simple command-line calculator program. The program should handle both integers and floating-point numbers.

(a) Write program that takes a mathematical expression as input and evaluates it. The program should support addition, subtraction, multiplication, and division.

(b) Extend the calculator to handle parentheses in mathematical expressions. Implement the necessary logic to evaluate expressions with nested parentheses.

(c) Enhance your calculator to include additional mathematical operations such as exponentiation, square root, and modulus.

(d) Implement error handling in the calculator program. Handle cases such as division by zero and invalid input expressions. Provide meaningful error messages.

(e) Implement a command to exit the calculator.

In [14]:
import math

#a) Function to evaluate mathematical expressions
def evaluate_expression(expression):
    try:
        # b)Evaluate the expression using eval(), which handles basic operations and parentheses
        result = eval(expression)
        return result
    # (d) Enhanced error handling
    except ZeroDivisionError:
        return "Error: Division by zero is not allowed."
    except SyntaxError:
        return "Error: Invalid syntax in the expression."
    except NameError:
        return "Error: Invalid input. Please enter valid numbers."
    except Exception as e:
        return f"Error: {e}"

def main():
    print("Welcome to the Advanced Calculator!")
    print("Supported operations: addition (+), subtraction (-), multiplication (*), division (/), exponentiation (**), modulus (%), square root (sqrt()), and parentheses ().")
    print("Type 'exit' to quit the calculator.\n")

    while True:
        # a)Prompt the user for a mathematical expression
        expression = input("Enter a mathematical expression (or type 'exit' to quit): ")

        #e) Check if the user wants to exit
        if expression.lower() == 'exit':
            print("Exiting the calculator. Goodbye!")
            break

        # c) Check for square root operation
        if 'sqrt' in expression:
            try:
                # Replace 'sqrt()' with math.sqrt() for handling square root
                expression = expression.replace('sqrt(', 'math.sqrt(')
            except Exception as e:
                print(f"Error: {e}")
                continue

        # Evaluate the expression and print the result
        result = evaluate_expression(expression)
        print(f"Result: {result}\n")

if __name__ == "__main__":
    main()


Welcome to the Advanced Calculator!
Supported operations: addition (+), subtraction (-), multiplication (*), division (/), exponentiation (**), modulus (%), square root (sqrt()), and parentheses ().
Type 'exit' to quit the calculator.

Enter a mathematical expression (or type 'exit' to quit): 5*3+2
Result: 17

Enter a mathematical expression (or type 'exit' to quit): (5*3)+2
Result: 17

Enter a mathematical expression (or type 'exit' to quit): ((5*3)+2)/2
Result: 8.5

Enter a mathematical expression (or type 'exit' to quit): ((5*3)+2)%2
Result: 1

Enter a mathematical expression (or type 'exit' to quit): sqrt(5*7)+4
Result: 9.916079783099615

Enter a mathematical expression (or type 'exit' to quit): sqrt(5*9)+4
Result: 10.70820393249937

Enter a mathematical expression (or type 'exit' to quit): sqrt((5*9)+4)
Result: 7.0

Enter a mathematical expression (or type 'exit' to quit): 9+3-12
Result: 0

Enter a mathematical expression (or type 'exit' to quit): exit
Exiting the calculator. Good

# Question 5 (30 points).

In the field of Data Science, data is often formatted as a comma-delimited (CSV) file, where each line in a file corresponds to a field's value. Refer to Chapter 9 (p. 106) for more information on comma-delimited files.

To answer this question, you will need to download the file [Assignment1_denton_housing.csv](https://raw.githubusercontent.com/unt-iialab/info5731-spring2022/main/assignments/Assignment1_denton_housing.csv) included with this assignment. The file contains statistics about housing in the city of Denton from 2008 to 2014. Write a program to implement the following questions.

(a) (5 pts) Calculate the difference in the number of Occupied Housing Units from year to year and print it. The difference must be calculated for the consecutive years such as 2008-2009, 2009-2010 etc. Finally, print the values in the ascending order.

(b) (5 pts) For all the years, calculate the percentage of housing units which are vacant an occupied. Print the results in the following format:

Year Vacant Housing Units Occupied Housing Units

    2008  30%   70%                          
	2009 ----- -----
	2010 ----- -----
    2011 ----- -----
	2012 ----- -----
    2013 ----- -----
	2014 ----- -----
  
  

(c) (10 pts) Calculate and print the valued and years in which the highest number of housing units were vacant and occupied. Print the results in the following format:  

                              Year	Value
	Vacant Housing Units	  -----   -----
	Occupied Housing Units	-----   -----

(d) (10 pts) Calculate the harmonic mean of the total housing units and print it out. You can find the information about harmonic mean here: https://ncalculators.com/statistics/harmonic-mean-calculator.htm

In [15]:
import pandas as pd
import statistics

# a) Function to compute year-over-year differences in occupied housing units
def compute_yearly_differences(data):
    occupied_data = data[data['title_field'] == 'Occupied Housing Units']
    occupied_data = occupied_data[['year', 'value']].sort_values(by='year')
    differences = occupied_data['value'].diff().dropna()
    sorted_differences = differences.sort_values()

    print("\nSorted Yearly Differences in Occupied Housing Units:")
    for i, diff in enumerate(sorted_differences, start=1):
        print(f"{occupied_data.iloc[i-1]['year']}-{occupied_data.iloc[i]['year']}: {diff}")
    return sorted_differences

# b) Function to calculate and display the percentage of vacant vs. occupied housing units
def calculate_unit_percentages(data):
    years = data['year'].unique()
    print(f"\n{'Year':<10} {'Vacant Units (%)':<20} {'Occupied Units (%)':<25}")
    for year in years:
        total_units = data[data['year'] == year]['value'].sum()
        vacant_units = data[(data['year'] == year) & (data['title_field'] == 'Vacant Housing Units')]['value'].values[0]
        occupied_units = data[(data['year'] == year) & (data['title_field'] == 'Occupied Housing Units')]['value'].values[0]
        vacant_pct = (vacant_units / total_units) * 100
        occupied_pct = (occupied_units / total_units) * 100
        print(f"{year:<10} {vacant_pct:<20.2f} {occupied_pct:<25.2f}")

# c) Function to find and print the year with maximum vacant and occupied units
def max_vacant_occupied_year(data):
    vacant_data = data[data['title_field'] == 'Vacant Housing Units']
    occupied_data = data[data['title_field'] == 'Occupied Housing Units']
    max_vacant_year = vacant_data.loc[vacant_data['value'].idxmax()]
    max_occupied_year = occupied_data.loc[occupied_data['value'].idxmax()]

    print("\nYear with Maximum Vacant and Occupied Units:")
    print(f"{'Category':<25} {'Year':<10} {'Units':<10}")
    print(f"Vacant Units               {max_vacant_year['year']:<10} {max_vacant_year['value']:<10}")
    print(f"Occupied Units             {max_occupied_year['year']:<10} {max_occupied_year['value']:<10}")

# d) Function to compute the harmonic mean of total housing units
def harmonic_mean_total_units(data):
    yearly_totals = data.groupby('year')['value'].sum()
    harmonic_mean = statistics.harmonic_mean(yearly_totals)
    print(f"\nHarmonic Mean of Total Housing Units: {harmonic_mean:.2f}")

# Main function to load data and call the relevant analysis functions
def main():
    # Loading data from the provided CSV file
    data = pd.read_csv('Assignment1_denton_housing.csv')

    # (a) Compute and print year-over-year differences in occupied housing units
    compute_yearly_differences(data)

    # (b) Calculate and display the percentage of vacant and occupied housing units by year
    calculate_unit_percentages(data)

    # (c) Find the year with the highest number of vacant and occupied housing units
    max_vacant_occupied_year(data)

    # (d) Compute the harmonic mean of total housing units over the years
    harmonic_mean_total_units(data)

if __name__ == "__main__":
    main()




Sorted Yearly Differences in Occupied Housing Units:
2008-2009: -795.0
2009-2010: -303.0
2010-2011: 1404.0
2011-2012: 1565.0
2012-2013: 2622.0
2013-2014: 5091.0

Year       Vacant Units (%)     Occupied Units (%)       
2014       1.94                 31.84                    
2013       3.00                 31.12                    
2012       2.26                 31.61                    
2011       3.21                 30.99                    
2010       2.77                 31.29                    
2009       3.10                 31.05                    
2008       3.74                 30.63                    

Year with Maximum Vacant and Occupied Units:
Category                  Year       Units     
Vacant Units               2008       4483      
Occupied Units             2014       46295     

Harmonic Mean of Total Housing Units: 130257.98


# Mandatory Question

Provide your thoughts on the assignment. What did you find challenging, and what aspects did you enjoy? Your opinion on the provided time to complete the assignment. Share any suggestions for improvement. Your feedback will be valuable for enhancing the learning experience.

In [16]:
'''
# Write your response below


Working with real-world data was made easier by this project, especially when it came to using Python to comprehend and manipulate CSV files.
The most difficult parts were correctly analyzing the data and using statistical techniques like the harmonic mean.
While challenging, correctly formatting the output was very satisfying.
I liked applying pandas and other Python modules to tackle real-world issues, which improved my data analytic abilities.
For those with some experience, the allotted time was sufficient; nevertheless, novices may want additional time.
More checkpoints and more explicit instructions on the dataset would enhance the experience and increase its educational value.
'''

'\n# Write your response below\n\n\nWorking with real-world data was made easier by this project, especially when it came to using Python to comprehend and manipulate CSV files. \nThe most difficult parts were correctly analyzing the data and using statistical techniques like the harmonic mean. \nWhile challenging, correctly formatting the output was very satisfying. \nI liked applying pandas and other Python modules to tackle real-world issues, which improved my data analytic abilities.\nFor those with some experience, the allotted time was sufficient; nevertheless, novices may want additional time.\nMore checkpoints and more explicit instructions on the dataset would enhance the experience and increase its educational value.\n'